[ImageCapture] Reject promises synchronously when track has ended

eric-carlson opened the pull request:
#### b0a7f8b8febb1740572fb77bf5cc615d30d34b50
<pre>
[ImageCapture] Reject promises synchronously when track has ended
<a href="https://bugs.webkit.org/show_bug.cgi?id=266297">https://bugs.webkit.org/show_bug.cgi?id=266297</a>
<a href="https://rdar.apple.com/119564866">rdar://119564866</a>

Reviewed by Youenn Fablet.

Reject takePhoto, getPhotoCapabilities, and getPhotoSettings promises synchronously
when called after a track has ended to match the spec.

* LayoutTests/fast/mediastream/image-capture-get-photo-capabilities-expected.txt:
* LayoutTests/fast/mediastream/image-capture-get-photo-capabilities.html:
* LayoutTests/fast/mediastream/image-capture-get-photo-settings-expected.txt:
* LayoutTests/fast/mediastream/image-capture-get-photo-settings.html:
* LayoutTests/fast/mediastream/image-capture-take-photo-expected.txt:
* LayoutTests/fast/mediastream/image-capture-take-photo.html:
* Source/WebCore/Modules/mediastream/ImageCapture.cpp:
(WebCore::ImageCapture::takePhoto):
(WebCore::ImageCapture::getPhotoCapabilities):
(WebCore::ImageCapture::getPhotoSettings):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
(WebCore::MediaStreamTrack::takePhoto):
(WebCore::MediaStreamTrack::getPhotoCapabilities):
(WebCore::MediaStreamTrack::getPhotoSettings):

Canonical link: <a href="https://commits.webkit.org/274437@main">https://commits.webkit.org/274437@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a575b8389b2fb9bbfd1c37e790dbdd4446dc9fed

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39003 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17936 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41356 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41537 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/34720 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41309 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20811 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15284 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/32652 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39577 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15106 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/33817 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13120 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13076 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34746 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42814 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35406 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35073 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38916 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13815 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11402 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37141 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15421 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/34025 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8747 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/15082 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14907 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on LayoutTests/fast/mediastream/image-capture-get-photo-capabilities.html:
```diff
@@ -20,9 +20,21 @@
+                assert_equals(result['name'], 'InvalidStateError');
+            });
+
+            await Promise.resolve();
```
It would read better to write it as:
```
await Promise.resolve();
assert_equals(result['name'], 'InvalidStateError');
return promise;
```
Ditto for the other tests.

eric-carlson committed:
[ImageCapture] Reject promises synchronously when track has ended
https://bugs.webkit.org/show_bug.cgi?id=266297
rdar://119564866

Reviewed by Youenn Fablet.

Reject takePhoto, getPhotoCapabilities, and getPhotoSettings promises synchronously
when called after a track has ended to match the spec.

* LayoutTests/fast/mediastream/image-capture-get-photo-capabilities-expected.txt:
* LayoutTests/fast/mediastream/image-capture-get-photo-capabilities.html:
* LayoutTests/fast/mediastream/image-capture-get-photo-settings-expected.txt:
* LayoutTests/fast/mediastream/image-capture-get-photo-settings.html:
* LayoutTests/fast/mediastream/image-capture-take-photo-expected.txt:
* LayoutTests/fast/mediastream/image-capture-take-photo.html:
* Source/WebCore/Modules/mediastream/ImageCapture.cpp:
(WebCore::ImageCapture::takePhoto):
(WebCore::ImageCapture::getPhotoCapabilities):
(WebCore::ImageCapture::getPhotoSettings):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
(WebCore::MediaStreamTrack::takePhoto):
(WebCore::MediaStreamTrack::getPhotoCapabilities):
(WebCore::MediaStreamTrack::getPhotoSettings):

Canonical link: https://commits.webkit.org/274437@main
